File: docker/node/execute.js
const fs = require('fs');
const { spawn } = require('child_process');

let code_and_input = '';
process.stdin.on('data', function (chunk) {
    code_and_input += chunk;
});

process.stdin.on('end', function () {
    const delimiter = '---SPLIT---\n';
    const parts = code_and_input.split(delimiter);

    const code = parts[0];
    const input_data = parts[1] || '';

    fs.writeFileSync('program.js', code);

    const nodeProcess = spawn('node', ['program.js']);

    nodeProcess.stdin.write(input_data);
    nodeProcess.stdin.end();

    nodeProcess.stdout.on('data', (data) => {
        process.stdout.write(data);
    });

    nodeProcess.stderr.on('data', (data) => {
        process.stderr.write(data);
    });

    nodeProcess.on('close', (code) => {
        process.exit(code);
    });
});

File: docker/perl/execute.pl
#!/usr/bin/perl
use strict;
use warnings;

my @code_and_input = <STDIN>;

my $delimiter = "---SPLIT---\n";
my $split_index = 0;
for my $i (0..$#code_and_input) {
    if ($code_and_input[$i] eq $delimiter) {
        $split_index = $i;
        last;
    }
}

my @code = @code_and_input[0..$split_index-1];
my @input_data = @code_and_input[$split_index+1..$#code_and_input];

my $input_file = "/tmp/input.txt";
open(my $input_fh, '>', $input_file) or die "Could not open input file: $!";
print $input_fh @input_data;
close($input_fh);

open(STDIN, '<', $input_file) or die "Could not redirect STDIN: $!";

my $code_str = join('', @code);
eval $code_str;
if ($@) {
    print STDERR $@;
    exit 1;
}
